### PR TITLE
Adding support for Gradient.ai

### PR DESCRIPTION
**Pull request by michaelfeil**
This PR adds support for the API at gradient.ai

Questions:
- Gradient.ai has no "hardcoded" model names, each model has a unique ID that can be finetuned. Is the implementation correct?

Feedback: (just some thoughts, feel free to ignore)
- It took a bit of time to integrate a new LLM, it could be improved by making the code in "main.py" less redundant. Maybe introduce a "BaseClass" to remove redundancy?
- Testing: I tried to write a test, however i found that none of the other llm providers did that. In langchain, there is the idea of "mock" the "request" library, to test APIs, additionally to integration tests.
- Local ci / pre-commit would  be of large use for: `mypy`, `import sorting`, `ruff`, `pytest`

### CODE REVIEW

**Comment by ghost** [comment]
Hey @michaelfeil, Thank you for the PR! 

- We have testing for each llm. it's in litellm/litellm/tests, there's multiple tests - test_completions.py, test_streaming.py, etc. 
- Acknowledging the feedback on base class. 
- We already have mypy + pytest for ci. Is there something i'm missing here? open to ideas.

**Comment by michaelfeil** [comment]
For 3. In addition to the CI, I would reproduce some parts, e.g. mypy, black, isort locally using a pre-commit. 
https://pre-commit.com/

**Comment by ishaan-jaff** [comment]
@michaelfeil 
- can you elaborate in what we should add in the `BaseClass` 
- which sections felt redundant that we can add to BaseClass

**Comment by ishaan-jaff** [comment]
tracking the pre-commit issue here: https://github.com/BerriAI/litellm/issues/422

**Comment by ishaan-jaff** [comment]
hi @michaelfeil sorry for the delay in reviewing this PR - we will review this in the next 24 hrs

File: docs/my-website/docs/providers/gradient.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# GRADIENT 
+
+LiteLLM supports from [Gradient](https://www.gradient.ai).
+
+### API KEYS
+You can set the API keys in the following ways:
+
+```python
+import os 
+os.environ["GRADIENT_ACCESS_TOKEN"] = ""
```
Comment (ishaan-jaff) [inline docs/my-website/docs/providers/gradient.md:10]: this needs to follow our standard format, it should be `GRADIENT_API_KEY`

File: docs/my-website/docs/providers/gradient.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+# set the secrets from https://auth.gradient.ai/select-workspace
+os.environ["GRADIENT_ACCESS_TOKEN"] = "DUMMY"
+os.environ["GRADIENT_WORKSPACE_ID"] = "821893_dummy_workspace"
+
+response = litellm.completion(
+    # set model to the ID (base or fine-tuned adapter) in gradient.ai
+    model="cc2dafce-9e6e-4a23-a918-cad6ba89e42e_base_ml_model",
+    custom_llm_provider='gradient',
+    messages=[
```
Comment (ishaan-jaff) [inline docs/my-website/docs/providers/gradient.md:34]: where moving towards standardizing to add custom_llm_provider as a prefix on `model`. Can you just set it such that `model="gradient/cc2dafce-9e6e-4a23-a918-cad6ba89e42e_base_ml_model"`

thoughts on this DX ?

File: litellm/llms/gradient.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        api_key=api_key,
+        additional_args={"complete_input_dict": data},
+    )
+    ## COMPLETION CALL
+    response = requests.post(
+        completion_url,
+        headers=headers,
+        data=json.dumps(data),
+    )
+    if "stream" in optional_params and optional_params["stream"] == True:
```
Comment (ishaan-jaff) [inline litellm/llms/gradient.py:112]: do they not support streaming?  if so can you mention that in gradient.md

File: litellm/main.py [hunk 4]
```diff
@@ -651,6 +652,38 @@ def completion(
...
+                print_verbose=print_verbose,
+                optional_params=optional_params,
+                litellm_params=litellm_params,
+                logger_fn=logger_fn,
+                encoding=encoding, 
+                api_key=gradient_key, 
+                logging_obj=logging,
+                gradient_workspace_id=gradient_workspace_id
+            )
+            if inspect.isgenerator(model_response) or ("stream" in optional_params and optional_params["stream"] == True):
```
Comment (ishaan-jaff) [inline litellm/main.py:680]: for providers that don't support streaming we usually fake streaming. Can you implement it in this PR please ? 

For all our providers we need to support regular completion calls + streaming(if they don't do streaming then we should fake streaming) + Exception mapping

Comment (ishaan-jaff) [reply litellm/main.py:680]: you can check out the Fake streaming implementation we did for sagemaker & palm for this

File: litellm/utils.py [hunk 5]
```diff
@@ -1279,28 +1296,30 @@ def load_test_model(
         }
 
 def validate_environment():
-        api_key = None
```
Comment (ishaan-jaff) [inline litellm/utils.py:1282]: why the large deletion here ? can you undo this

File: litellm/utils.py [hunk 6]
```diff
@@ -2695,6 +2714,8 @@ def __next__(self):
                 elif self.custom_llm_provider and self.custom_llm_provider == "ai21": #ai21 doesn't provide streaming
                     chunk = next(self.completion_stream)
                     completion_obj["content"] = self.handle_ai21_chunk(chunk)
+                elif self.custom_llm_provider and self.custom_llm_provider == "gradient": #gradient doesn't provide streaming
+                    pass
```
Comment (ishaan-jaff) [inline litellm/utils.py:2718]: we need fake streaming here

File: litellm/utils.py [hunk 7]
```diff
@@ -974,6 +974,14 @@ def get_optional_params(  # use the openai defaults
         optional_params["num_beams"] = num_beams
         if max_tokens != float("inf"):
             optional_params["max_new_tokens"] = max_tokens
+    elif custom_llm_provider == "gradient":
+        optional_params["temperature"] = temperature
```
Comment (ishaan-jaff) [inline litellm/utils.py:978]: don't you have to check if temperature != inf and only in that case set it

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
thanks for the awesome work on this! requested some changes

**Comment by ghost** [comment]
@michaelfeil any updates on this?

**Comment by ghost** [comment]
closing as there are no updates here
